### PR TITLE
expression: rewrite builtin function: TIME

### PR DESCRIPTION
**Pull request by winkyao**
for #4080

### CODE REVIEW

File: expression/integration_test.go [hunk 1]
```diff
@@ -931,6 +931,24 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
...
+	result = tk.MustQuery("select time('01:02:03')")
+	result.Check(testkit.Rows("01:02:03"))
+	result = tk.MustQuery("select time('-838:59:59.000000')")
+	result.Check(testkit.Rows("-838:59:59.000000"))
+	result = tk.MustQuery("select time('-838:59:59.000001')")
+	result.Check(testkit.Rows("-838:59:59.000000"))
+	result = tk.MustQuery("select time('-839:59:59.000000')")
+	result.Check(testkit.Rows("-838:59:59.000000"))
+	result = tk.MustQuery("select time('840:59:59.000000')")
+	result.Check(testkit.Rows("838:59:59.000000"))
```
Comment (zz-jason) [inline expression/integration_test.go]: add test for invalid strings and null input

File: util/types/fsp.go [hunk 2]
```diff
@@ -33,7 +33,8 @@ const (
 	DefaultFsp int = 0
 )
 
-func checkFsp(fsp int) (int, error) {
+// CheckFsp check fsp is in valid range.
```
Comment (zz-jason) [inline util/types/fsp.go]: s/check/checks/

File: expression/integration_test.go (hunk 1 above)
Comment (winkyao) [inline expression/integration_test.go]: The boundary condition is not handled correctlly by TiDB, see #4193 , it is another issue, should solve it in another pr.

File: util/types/fsp.go (hunk 2 above)
Comment (winkyao) [inline util/types/fsp.go]: Done! @zz-jason

File: expression/builtin_time.go [hunk 3]
```diff
@@ -1414,41 +1414,41 @@ func (c *timeFunctionClass) getFunction(args []Expression, ctx context.Context)
...
 	}
 
-	str, err := args[0].ToString()
-	if err != nil {
-		return d, errors.Trace(err)
-	}
-	idx := strings.Index(str, ".")
 	fsp := 0
+	idx := strings.Index(expr, ".")
 	if idx != -1 {
```
Comment (zz-jason) [inline expression/builtin_time.go]: maybe this way is better:

```go
if idx := strings.Index(expr, "."); idx != -1 {
```

rest LGTM

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL.

File: util/types/fsp.go [hunk 4]
```diff
@@ -33,7 +33,8 @@ const (
 	DefaultFsp int = 0
 )
 
-func checkFsp(fsp int) (int, error) {
+// CheckFsp checks fsp is in valid range.
```
Comment (XuHuaiyu) [inline util/types/fsp.go]: ... whether fsp is ...

Comment (winkyao) [inline util/types/fsp.go]: Done

**Pull request merged**